File: src/main/java/com/lessmarkup/interfaces/module/ModuleInitializer.java
package com.lessmarkup.interfaces.module;

import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.data.Migration;
import com.lessmarkup.interfaces.recordmodel.RecordModel;
import com.lessmarkup.interfaces.structure.NodeHandler;
import java.util.Collection;

public interface ModuleInitializer {
    default void initialize() {}
    String getName();
    String getModuleType();
    Collection<Class<? extends RecordModel>> getModelTypes();
    Collection<Class<? extends DataObject>> getDataObjectTypes();
    Collection<Class<? extends NodeHandler>> getNodeHandlerTypes();
    Collection<Class<? extends Migration>> getMigrations();
}

File: src/main/scala/com/lessmarkup/interfaces/module/ModuleInitializer.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.module

import com.lessmarkup.interfaces.data.{DataObject, Migration}
import com.lessmarkup.interfaces.recordmodel.RecordModel
import com.lessmarkup.interfaces.structure.NodeHandlerFactory

trait ModuleInitializer {
  def initialize() {
  }

  def getName: String
  def getModuleType: String
  def getModelTypes: Seq[Class[_ <: RecordModel[_]]]
  def getDataObjectTypes: Seq[Class[_ <: DataObject]]
  def getNodeHandlerTypes: Seq[Class[_ <: NodeHandlerFactory]]
  def getMigrations: Seq[Class[_ <: Migration]]
}